cp1616/src/cp1616_io_controller.cpp

```diff
@@ -0,0 +1,510 @@
+    in_data_        = new PNIO_UINT8* [device_input_count_ * sizeof(PNIO_UINT8)];
+
+    //Set initial InpData values to zero
+    memset(in_module_data_, 0, total_input_size_);
+
+    unsigned int i, j;
+
+    //Assign array pointers
+    for(i = 0, j = 0; j < device_input_count_; )
+    {
+      in_data_[j] = &(in_module_data_[i]);
+      i += device_input_length_[j++];
+    }
+
+    //Print InData array information
+    ROS_INFO_STREAM("Input Data Array: [Total size: " << total_input_size_ << " bytes]");
+    for(i = 0; i < device_input_count_; i++)
+    {
+      if((device_input_length_[i]-1) == 0)
+      {
+        ROS_INFO_STREAM("I: "<< device_input_address_[i].u.Addr
+          << " - " << device_input_address_[i].u.Addr + device_input_length_[i] - 1
+          << ": InData  "  << "[" << i << "]" << "[0]");
+      }
+      else
+      {
+        ROS_INFO_STREAM("I: "<< device_input_address_[i].u.Addr
+          << " - " << device_input_address_[i].u.Addr + device_input_length_[i] - 1
+          << ": InData  "  << "[" << i << "]" << "[0] ... "
+          << "[" << i << "]" << "[" << device_input_length_[i]-1 << "]");
+      }
+    }
+  }
+
+  //-------------------------------------------------------------------------
+  //Allocate memory for registered output modules
+  //-------------------------------------------------------------------------
+  if(device_output_count_ > 0)    //if any output module available
+  {
+    //Memory allocation
+    out_module_data_ = new PNIO_UINT8 [total_output_size_ * sizeof(PNIO_UINT8)];
+    out_data_        = new PNIO_UINT8* [device_output_count_ * sizeof(PNIO_UINT8)];
+
+    //Set initial InpData values to zero
+    memset(out_module_data_, 0, total_output_size_);
+
+    unsigned int i, j;
+
+    //Assign array pointers
+    for(i = 0, j = 0; j < device_output_count_; )
+    {
+      out_data_[j] = &(out_module_data_[i]);
+      i += device_output_length_[j++];
+    }
+
+    //Print OutData array information
+    ROS_INFO_STREAM("Output Data Array: [Total size: " << total_output_size_ << " bytes]");
+    for(i = 0; i < device_output_count_; i++)
+    {
+      if((device_output_length_[i]-1) == 0)
+      {
+        ROS_INFO_STREAM("I: "<< device_output_address_[i].u.Addr
+          << " - " << device_output_address_[i].u.Addr + device_output_length_[i] - 1
+          << ": OutData "  << "[" << i << "]" << "[0]");
+      }
+      else
+      {
+         ROS_INFO_STREAM("I: " << device_output_address_[i].u.Addr
+           << " - " << device_output_address_[i].u.Addr + device_output_length_[i] - 1
+           << ": OutData "  << "[" << i << "]" << "[0] ... "
+           << "[" << i << "]" << "[" << device_output_length_[i]-1 << "]");
+      }
+    }
+  }
+}
+
+int Cp1616IOController::addInputModule(unsigned int input_size, unsigned int input_start_address)
+{
+  if(cp_current_mode_ == PNIO_MODE_OPERATE)
+  {
+    ROS_ERROR_STREAM("Not able to add Input module in Operate state!");
+    return -1;
```

Comment:
IO Base library includes file [pnioerrx.h] (http://www.smartroboticsys.eu/wp-content/uploads/2015/07/pnioerrx.h) - list of all potential error codes that is included in both controller/device headers . My intention is to return these error-codes coming from IO Base functions back to the user. You are right there were 3 or 4 returns with -1 or 0 values in existing code, I have already assigned them values from **pnioerrx.h**